airflow/models/dag.py

```diff
@@ -533,94 +537,54 @@ def next_dagrun_info(
             "automated" DagRuns for this dag (scheduled or backfill, but not
             manual)
         """
-        if (
-            self.schedule_interval == "@once" and date_last_automated_dagrun
-        ) or self.schedule_interval is None:
-            # Manual trigger, or already created the run for @once, can short circuit
+        # XXX: The timezone.coerce_datetime calls in this function should not
+        # be necessary since the function annotation suggests it only accepts
+        # pendulum.DateTime, and someone is passing datetime.datetime into this
+        # function. We should fix whatever is doing that.
+        if self.is_subdag:
             return (None, None)
-        next_execution_date = self.next_dagrun_after_date(date_last_automated_dagrun)
-
-        if next_execution_date is None:
+        time_table: TimeTable = self.time_table
+        restriction = self._format_time_restriction()
+        if not self.catchup:
+            restriction = time_table.cancel_catchup(restriction)
```

Comment:
`not` ... `cancel` is a bit hard for me to follow (akin to a double negative.)

How about this as an alternative name?

```suggestion
        if not self.catchup:
            restriction = time_table.skip_to_latest(restriction)
```

An additional thought: this should probably be handled _inside_ `time_table.next_dagrun_info`, so how about we change the signature of that to:


```python
    def next_dagrun_info(
        self,
        last_automated_dagrun: Optional[DateTime],
        between: TimeRestriction,
        catchup: boolean,
    ) -> Optional[DagRunInfo]:
```

I think by passing in the catchup and letting the Timetable handle it that also means we don't have to have `cancel_catchup` as part of the protocol.

Reply:
The thing I don’t particular like about this is `catchup` does not make sense for any runs after the first, which to me indicates the design isn’t right. If we want to make catchup handling a part of the time table, a better interface to me would be to separate the first and subsequent DAG runs:

```python
def first_dagrun_info(
    self,
    between: TimeRestriction,
    catchup: bool,
) -> Optional[DagRunInfo]:
    ...

def next_dagrun_info(
    self,
    last_automated_dagrun: DateTime,
    between: TimeRestriction,
) -> Optional[DagRunInfo]:
    ...
```

Reply:
Ah -- catch can  have an effect later on too.

If I pause a dag with a daily schedule for 5 days, and then resume it:

- With catchup: missing dag runs will be created
- Without catchup: only the "latest" dag run will be created.

Reply:
Only your idea would work then. Cool, I’ll change it.

Reply:
It also makes me wonder maybe I had implemented something wrong to not have realised this. I’ll need to take another look at my changes.

Reply:
I decided to put the catchup argument in `TimeRestriction` instead (since that’s a time restriction as well, basically move the earliest possible time to the current time).